### ISSUE TITLE
Add `numba` and `llvmlite` `musl` wheels for `alpine` support

### ISSUE THREAD

**Issue by mpierrau** [numba/llvmlite#1034]
## Feature request

Hi, I was wondering if it is possible to add [MUSL wheels](https://musl.libc.org/) for `numba` and `llvmlite` since lots of people use CICD images based on [Alpine](https://alpinelinux.org/). This distribution uses different a implementation of libc: [musl](https://musl.libc.org/) (see [Alpine Linux has switched to musl libc | Alpine Linux ](https://alpinelinux.org/posts/Alpine-Linux-has-switched-to-musl-libc.html)), and is incompatible with `manylinux` wheels.

I know that for example the `shap` library is [waiting for `llvmlite` and `numba` to release MUSL wheels](https://github.com/shap/shap/pull/3087).

Thanks!

**Comment by esc** [comment]
@mpierrau quick question about this: where are they distributed, can they be uploaded to PyPi too? How many artifacts are we talking about here?

**Comment by esc** [comment]
@mpierrau we just discussed this during our issue triage and it is unlikely that this will happen anytime soon due to a maintainer resource shortage. Would suggest to leave this open anyway, in case that resources become available over the coming months or years.

**Comment by mpierrau** [comment]
I have no experience from distributing Python packages, but from what I can gather, the `musl` wheels are also uploaded to `PyPi`, and it's one wheel artifact per architecture you wish to support. The number of architectures supported for `musl` are in general the same as for `manylinux` (from looking in my lockfiles for other packages which do distribute `musl` wheels to `PyPi`).

It looks like [`cibuildwheel`](https://cibuildwheel.readthedocs.io/en/stable/) supports building `musl` wheels for `CPython` 3.6 - 3.12, for the following architectures `x86_64`, `aarch64`, `i686`, `s390x`, `ppc64le`.

**Comment by esc** [comment]
@mpierrau thank you for digging into this, much appreciated.

Thoughts:

a) We have space issues on PyPi with llvmlite. Our wheels contain LLVM itself and we regularly run out of space on PyPi because of this. Adding another 4 (or 8?) wheels per release may increase the size requirement so we'll probably need additional checks before uploading, if there is enough space left. This isn't something that many or even most Python packages face so tooling here -- for example to check for space -- is very rudimentary and sometimes not reliable. We usually only notice that we run out of space when a release fails to upload and then need to request more space on PyPi which can sometimes take several days to process from the PyPi side.

b) As you have noticed, llvmlite is quite unique in how it must be built, hence we don't (and maybe even can't) use cibuildwheel but instead have a very custom solution consisting of several bare metal machines that we maintain ourselves. Thus, adding musl wheels will require significantly more custom tooling and work than for an average package. It's not known how long doing this will take and may include multiple rabbit holes. I would estimate it takes between 3-4 weeks of developer time, so this is in the "nice to have" category for me as you have been the only person asking about this so far.

Of course, this could receive a higher priority if there is a significant ask (hair on fire need) for this from the community.

**Comment by mpierrau** [comment]
@esc Thanks for the reply and the insight to your situation. It is indeed a "nice to have", and my hair currently isn't on fire, so I'll patiently wait and come back to you with a bump if it spontaneously ignites in the future :)

**Comment by esc** [comment]
> @esc Thanks for the reply and the insight to your situation. It is indeed a "nice to have", and my hair currently isn't on fire, so I'll patiently wait and come back to you with a bump if it spontaneously ignites in the future :)

🔥🔥🔥

**Comment by esc** [comment]
> > @esc Thanks for the reply and the insight to your situation. It is indeed a "nice to have", and my hair currently isn't on fire, so I'll patiently wait and come back to you with a bump if it spontaneously ignites in the future :)
> 
> 🔥🔥🔥

No, but seriously, this is laborious and may come with challanges, but not technically impossible or not feasible. Maybe as musl wheels become more widely used, well see a need to have them for Numba too.

PS: do you know if NumPy provides these already?

**Comment by mpierrau** [comment]
> No, but seriously, this is laborious and may come with challanges, but not technically impossible or not feasible. Maybe as musl wheels become more widely used, well see a need to have them for Numba too.

My impression from lurking in some major package issue boards is that the requests for musl wheels are gaining traction, but I don't have the pulse on whether it's requested by your community.

> PS: do you know if NumPy provides these already?

NumPy started providing musl wheels in their [`1.25.0` release](https://pypi.org/project/numpy/1.25.0/#files) (June 2023), [this](https://github.com/numpy/numpy/issues/20089) is an issue where some pro's and con's were discussed, especially [this comment](https://github.com/numpy/numpy/issues/20089#issuecomment-1414412193) had some good points:

> > Is there enough use of musl to justify it?
> 
> I'd say yes. Musl is used by Alpine Linux, which is quite popular for use in Docker containers, because of the small-sized images it produces (thanks to Musl being so much smaller than Glibc). We get regular requests, probably more than for any other platform. And several Pandas and SciPy devs seem to want this.
> 
> > Can pip can distinguish the linked libraries?
> 
> Yes, there is a separate `musllinux` wheel spec.
> 
> > Does this affect OpenBLAS maintenance?
> 
> Only insofar as a separate build of OpenBLAS must be done for this. Which is available - hopefully maintenance of that will be smooth sailing.
> 
> > Do downstream projects also need to link with musl?
> 
> Yes - but a separate wheel tag and Linux distro takes care of that. However, what you may be getting at is that one needs a complete collection of `musllinux` wheels for all the packages one needs. Patchy support isn't very useful. So it'd be good to see at least Pandas, SciPy, scikit-learn, and Matplotlib wheels. My impression is that that will happen if NumPy leads the way.
> 
> > Does the MIT license add anything?
> 
> No worries there, all good on the licensing front.
> 
> > Could static linking be useful?
> 
> That's actually a common reason for using Musl too. It's not useful for NumPy I'd say, but in general static linking with Musl is a great way to produce binaries that run on any distro (because no dependency on the distro-shipped libc).
> 
> > At some point CI stability might become a problem.
> 
> That's always a worry, but more so for nonstandard hardware than for something that runs on plain Linux x86-64 and aarch64 instances. So for Musl it should be pretty good compared to our other candidates for inclusion in our zoo of wheels.

In addition, it looks like Pandas are also [considering it](https://github.com/pandas-dev/pandas/issues/50511), but are waiting for PyArrow, whom
[are working on it](https://github.com/apache/arrow/issues/18036).

I asked the same [question](https://github.com/scikit-learn/scikit-learn/issues/27004#issuecomment-1966926217) in scikit-learn and they replied that:
> For the record this was discussed at yesterday's monthly dev meeting and for now the consensus seems to be that the extra maintenance effort and CI costs are not necessarily worth it.
[...]
> the maintainers might reconsider that decision (not to provide musllinux wheels on pypi.org) based on expressed user feedback e.g. next year.

SciPy are supporting providing them since [this fall](https://github.com/scipy/scipy/issues/18981) and [so is matplotlib](https://github.com/matplotlib/matplotlib/issues/26438) (just to give you an idea).

**Comment by fahedouch** [comment]
could you enhance this ticket please, it is blocking us to manipulate  `llvm` in `alpine`

**Comment by esc** [comment]
> could you enhance this ticket please, it is blocking us to manipulate `llvm` in `alpine`

@fahedouch thank you for messaging us about this! Sorry to hear that a lack of alipe based packages is blocking you.

For reference, the recently established support policy for Numba and llvmlite is here:

https://numba.readthedocs.io/en/latest/reference/support_tiers.html?highlight=tier

I don't yet know yet if this will be accepted as a Tier 1 platform, but I will raise this during the developer meeting next Tuesday to see if it's possible to get this prioritized.

**Comment by esc** [comment]
@fahedouch if you would like to help out here, maybe you can check if the following pre-requisites are met:


* A conda based Python distribution for the operating system and hardware target is required. This is because the public CI/CD for the Numba stack is driven by conda and conda packages (even for wheel builds). The purpose of this condition is to reduce the overall CI/CD maintenance burden.
* The operating system and hardware target must be supported by GitHub Actions on the free-tier.
* The release package type must be conda or wheel.
* The distribution system must be PyPI or Anaconda repositories.

**Comment by esc** [comment]
@fahedouch I spoke to a few people and one open, glaring question stood out: would it be possible/feasible for Alpine-Linux to package the Numba/llvmlite stack such that you can just `apt install` them? Like, why do you need wheels?

**Comment by strophy** [comment]
@esc I'm an Alpine maintainer and wanted to add this package in order to enable tests on some other packages that had tests disabled due to unavailability of numba and llvmlite on musl. I'd appreciate any help and comments here: https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/107692
